### PR TITLE
prov/psm2&3: get cq data for PSMX3_IOV_RECV_CONTEXT

### LINKED ISSUE #7762: prov/psm3: tsendv losing cq data
**Describe the bug**
We are chasing a test failure and it appears that `psm3` is losing CQ data when another process is sending multiple IOV data with non-zero CQ data. We triggered this bug using `fi_tsendmsg` but I believe it applies to the `fi_tsendv` path. I chased the bug down to this line:
https://github.com/ofiwg/libfabric/blob/7d8421b76ff7be9ae9a79148c2805ec60faed3e9/prov/psm3/psm3/psm_mq_recv.c#L509
When messages are sent in multiple segments, the later segments (I believe) are overwriting the tag (including the CQ data) that is set in the initial envelope. I am not familiar enough with the codebase to attempt a fix.

**To Reproduce**
Current MPICH, configure with  `--with-device=ch4:ofi --with-libfabric=embedded --without-ch4-shmmods`, and run the test in `test/mpi/pt2pt/sendself` using command line:
```
mpirun -n 2 ./sendself -type=MPI_INT -sendcnt=8192 -recvcnt=8192 -seed=47 -testsize=100
```

I am a bit surprised only this test triggered the bug.

**Expected behavior**
A single output "No Errors"

**Output**
A error message about unmatched STATUS, and it hangs. We use CQ data for sender ranks, and it hangs in `MPI_Ssend` due to replying to the wrong sender due to missing CQ data.

@acgoldma

### PR DESCRIPTION
**Pull request by hzhou**
The CQ data was not retrieved in psmx3_mq_status_copy for the case of
PSMX3_IOV_RECV_CONTEXT, causing loss of CQ data in fi_tsendmsg.

Fixes #7762 

EDIT: Fixes applied to both psm2 and psm3.

### CODE REVIEW

File: prov/psm3/src/psmx3_cq.c
Comment (hzhou) [reply prov/psm3/src/psmx3_cq.c:768]: Oh, you mean the `if-`condition below. Yeah, that makes sense. I just imitated the previous code patterns. I'll let @acgoldma decide.

Comment (mwheinz) [reply prov/psm3/src/psmx3_cq.c:768]: @acgoldma asked me to take a look at this; I think @shefty is correct - move line 768 and put it between 769 and 770. (Adding the appropriate braces on line 769 and 770... ;-) )

Comment (hzhou) [reply prov/psm3/src/psmx3_cq.c:768]: That is fine. My question is, for consistency reasons, should I do that for all instances? --
https://github.com/ofiwg/libfabric/blob/9cc1cf74e5a886647e1846d58744c14fa63c4979/prov/psm3/src/psmx3_cq.c#L439-L441
https://github.com/ofiwg/libfabric/blob/9cc1cf74e5a886647e1846d58744c14fa63c4979/prov/psm3/src/psmx3_cq.c#L461-L463
https://github.com/ofiwg/libfabric/blob/9cc1cf74e5a886647e1846d58744c14fa63c4979/prov/psm3/src/psmx3_cq.c#L483-L485
https://github.com/ofiwg/libfabric/blob/9cc1cf74e5a886647e1846d58744c14fa63c4979/prov/psm3/src/psmx3_cq.c#L505-L507
https://github.com/ofiwg/libfabric/blob/9cc1cf74e5a886647e1846d58744c14fa63c4979/prov/psm3/src/psmx3_cq.c#L622-L624

Comment (mwheinz) [reply prov/psm3/src/psmx3_cq.c:768]: Ah. Sorry, I see your issue now. So, you're actually changing code that has persisted since at least when psm >2< was first developed, and I'm concerned that changing it in all such places might have more side-effects than either of us realize, but I'm also concerned that it could end up getting set to a non-zero value (in the change you want to make) which might also have side-effects.

Which MPI are you running, or are you using libfabric directly?

Comment (hzhou) [reply prov/psm3/src/psmx3_cq.c:768]: We are MPICH developers. MPICH is using libfabric directly.

Reading the man page on `fi_cq`, it does say that we are supposed ignore `data` if `FI_REMOTE_CQ_DATA` is not set. However, in our current code we don't actually check the completion flag, but always assumed the data is there. Part of the reason is for the path we use CQ data, we kind of depend on CQ data being sent. It'll be an error anyway if we send CQ data but couldn't receive it. Therefore, I believe move the data assignment into the `if-`branches to ensure consistency should be generally preferred. I'll add to the commit if you give me the go-ahead.

> I'm concerned that changing it in all such places might have more side-effects than either of us realize ...

Can we trust the tests? :)

Comment (hzhou) [reply prov/psm3/src/psmx3_cq.c:768]: We just switched to `psm3` in our testing a couple of weeks ago. Before that, we are using `sockets`. I'll check with Intel (Texas) team. They routinely test on `psm2` and I haven't heard any issue from them.

Comment (ToddRimmer) [reply prov/psm3/src/psmx3_cq.c:768]: [@hzhou](https://github.com/hzhou) - you may also want to look at OpenMPI and check with some of the Intel MPICH experts on how this is handled by those MPIs.  PSM3 support for OFI is very similar to PSM2 and both OpenMPI and Intel MPI have successfully worked over PSM2 for over 10 years, and more recently have been successfully running over PSM3 for the last few years.

Comment (shefty) [reply prov/psm3/src/psmx3_cq.c:768]: Other MPIs may not be using CQ data.

Comment (hzhou) [reply prov/psm3/src/psmx3_cq.c:768]: I'll add a patch to `psm2` later

File: prov/psm3/src/psmx3_cq.c
Comment (hzhou) [inline prov/psm3/src/psmx3_cq.c:769]: Hmm, I guess I should replace `PSMX3_STATUS_TAG(req)` with `sendv_rep->tag`

File: prov/psm3/src/psmx3_cq.c
Comment (hzhou) [reply prov/psm3/src/psmx3_cq.c:768]: The question of whether to set `data` inside the branch was introduced here: https://github.com/ofiwg/libfabric/pull/5863. From the comment, it overloads CQ data for the purpose of `sep_id` for SEP address lookup optimization. Thus, the pattern was on purpose.

Comment (hzhou) [reply prov/psm3/src/psmx3_cq.c:768]: @j-xiong Please review the `psm2` fix.

**Review by j-xiong — COMMENTED** [review]
Both psm2 and psm3 fixes look fine to me. 

+@timothom64 (current psm2 owner) to double check.

File: prov/psm3/src/psmx3_cq.c
Comment (shefty) [reply prov/psm3/src/psmx3_cq.c:768]: @mwheinz - Are the changes here okay?

File: prov/psm2/src/psmx2_cq.c
Comment (BrendanCunningham) [inline prov/psm2/src/psmx2_cq.c:777]: On the send side, in both `psmx2_sendv_generic()` and `psmx2_tagged_sendv_generic()`, when `req->iov_protocol == PSMX2_IOV_PROTO_MULTI`, the send-tag is constructed with neither scalable endpoint ID nor CQ data and the  `PSMX2_IMM_BIT` is not set in this case. I don't know if there's anything the receiver has to do in this case.

**Comment by BrendanCunningham** [comment]
@hzhou I see the PSM3 patch for this is merged into the pmodels/libfabric fork used by pmodels/mpich libfabric submodule; do you intend to merge the PSM2 commit in as well, either before or after it's merged into ofiwg/libfabric?

**Comment by hzhou** [comment]
> @hzhou I see the PSM3 patch for this is merged into the pmodels/libfabric fork used by pmodels/mpich libfabric submodule; do you intend to merge the PSM2 commit in as well, either before or after it's merged into ofiwg/libfabric?

Yes, we plan to merge the psm2 commit as well. Since our CI testing does not use `psm2`, it has a lesser urgency, so I am waiting for the conclusion here.

File: prov/psm2/src/psmx2_cq.c
Comment (BrendanCunningham) [inline prov/psm2/src/psmx2_cq.c:1513]: Lines 1513-1515: should use leading tabs instead of spaces.

**Review by BrendanCunningham — CHANGES_REQUESTED** [review]
PSM2 part looks good apart from leading-spaces-instead-of-tabs in one place.

File: prov/psm2/src/psmx2_cq.c
Comment (hzhou) [reply prov/psm2/src/psmx2_cq.c]: Thanks. Forget the tabs every single time... Fixed.

**Review by BrendanCunningham — APPROVED** [review]
PSM2 changes look good.

**Comment by j-xiong** [comment]
I plan to merge this PR if no objection is heard by the end of Thursday 7/7.

**Pull request merged**
